tests/integration/s2n_resume_test.py

```diff
@@ -0,0 +1,151 @@
+#
+# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#  http://aws.amazon.com/apache2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+#
+
+import os
+import sys
+import time
+import ssl
+import socket
+import subprocess
+
+# All supported ciphers
+S2N_CIPHERS= [
+    ("RC4-MD5", ssl.PROTOCOL_SSLv3),
+    ("RC4-SHA", ssl.PROTOCOL_SSLv3),
+    ("DES-CBC3-SHA", ssl.PROTOCOL_SSLv3),
+    ("EDH-RSA-DES-CBC3-SHA", ssl.PROTOCOL_SSLv3),
+    ("AES128-SHA", ssl.PROTOCOL_TLSv1),
+    ("DHE-RSA-AES128-SHA", ssl.PROTOCOL_TLSv1),
+    ("AES256-SHA", ssl.PROTOCOL_TLSv1),
+    ("DHE-RSA-AES256-SHA", ssl.PROTOCOL_TLSv1),
+    ("AES128-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("AES256-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("DHE-RSA-AES128-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("DHE-RSA-AES256-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("AES128-GCM-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("AES256-GCM-SHA384", ssl.PROTOCOL_TLSv1_2),
+    ("DHE-RSA-AES128-GCM-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("ECDHE-RSA-DES-CBC3-SHA", ssl.PROTOCOL_TLSv1),
+    ("ECDHE-RSA-AES128-SHA", ssl.PROTOCOL_TLSv1),
+    ("ECDHE-RSA-AES256-SHA", ssl.PROTOCOL_TLSv1),
+    ("ECDHE-RSA-AES128-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("ECDHE-RSA-AES256-SHA384", ssl.PROTOCOL_TLSv1_2),
+    ("ECDHE-RSA-AES128-GCM-SHA256", ssl.PROTOCOL_TLSv1_2),
+    ("ECDHE-RSA-AES256-GCM-SHA384", ssl.PROTOCOL_TLSv1_2),
+]
+
+PROTO_VERS_TO_STR = {
+    ssl.PROTOCOL_SSLv3 : "SSLv3",
+    ssl.PROTOCOL_TLSv1 : "TLSv1.0",
+    ssl.PROTOCOL_TLSv1_1 : "TLSv1.1",
+    ssl.PROTOCOL_TLSv1_2 : "TLSv1.2",
+}
+
+PROTO_VERS_TO_ARG = {
+    ssl.PROTOCOL_SSLv3 : "-ssl3",
+    ssl.PROTOCOL_TLSv1 : "-tls1",
+    ssl.PROTOCOL_TLSv1_1 : "-tls1_1",
+    ssl.PROTOCOL_TLSv1_2 : "-tls1_2",
+}
+
+def try_resume(endpoint, port, cipher, ssl_version):
+    # Fire up s2nd
+    s2nd = subprocess.Popen(["../../bin/s2nd", "-c", "test_all", str(endpoint), str(port)], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+
+    # Make sure it's running
+    s2nd.stdout.readline()
+
+    # Fire up s_client
+    s_client = subprocess.Popen(["../../libcrypto-root/bin/openssl", "s_client", PROTO_VERS_TO_ARG[ssl_version], "-cipher", cipher,
+                                 "-quiet", "-reconnect", "-connect", str(endpoint) + ":" + str(port)], stdin=subprocess.PIPE, stdout=subprocess.PIPE,
+                                 stderr=subprocess.DEVNULL)
+
+    # Wait until the 6th connection
+    seperators = 0
+    for line in s2nd.stdout:
+        line = line.decode("utf-8").strip()
+        if line.startswith("Cipher negotiated"):
```

Comment:
Does this verify that the session is being resumed? If we use -quiet s_client won't output the session details. We could change this for loop to detect the presence of the "Reused" word. We'd expect 1 "new" session output for the first connection attempt followed by 4 reuseds.

```
---
New, TLSv1/SSLv3, Cipher is AES128-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1
    Cipher    : AES128-SHA
    Session-ID: 14D43C09E3EBC9DC674265C9977F4DA2E04E74A486FAA7BBB9792E5C4ABF1CA4
    Session-ID-ctx:
    Master-Key: 9712687212887DEB6E3479FEA7B9367680349AE067466611684326AE1ED634CBFA1E1CB24AA1383D07BC78AB8D2B1413
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 100800 (seconds)
    TLS session ticket:
    0000 - 5d a2 d5 29 fe 27 11 4e-31 77 56 f6 d2 80 ab 53   ]..).'.N1wV....S
    0010 - 6c 2c e0 97 8e bf c0 07-dd 66 74 6a 70 c3 f7 ca   l,.......ftjp...
    0020 - aa a8 20 b7 0a 9e 05 69-97 b2 84 b4 d3 1d fa 38   .. ....i.......8
    0030 - 3b 7f 17 f7 2a 4e 2e 3e-cc 53 08 49 bc ca 9b c7   ;...*N.>.S.I....
    0040 - 08 46 ec b0 e8 1d 52 97-34 4d 1d 6f f4 2f 7f 86   .F....R.4M.o./..
    0050 - 10 b5 0b 4f 85 73 c6 4c-ed 4b c0 77 c8 c3 45 d1   ...O.s.L.K.w..E.
    0060 - d6 35 a6 62 0a 41 4f b3-5e 33 e0 65 ec b4 19 e4   .5.b.AO.^3.e....
    0070 - ce a2 b1 62 dd 29 8e 2c-33 8b 5e 67 f1 03 f5 b0   ...b.).,3.^g....
    0080 - 9d 4a 0b b2 30 10 83 4d-ae ac 66 20 41 99 6a bf   .J..0..M..f A.j.
    0090 - ed b0 b1 34 66 76 92 63-7d 63 f0 a4 e1 4b 5d 77   ...4fv.c}c...K]w
    00a0 - ec 24 ed ab                                       .$..

    Start Time: 1467238146
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
drop connection and then reconnect
CONNECTED(00000003)
---
Reused, TLSv1/SSLv3, Cipher is AES128-SHA
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1
    Cipher    : AES128-SHA
    Session-ID: 14D43C09E3EBC9DC674265C9977F4DA2E04E74A486FAA7BBB9792E5C4ABF1CA4
    Session-ID-ctx:
    Master-Key: 9712687212887DEB6E3479FEA7B9367680349AE067466611684326AE1ED634CBFA1E1CB24AA1383D07BC78AB8D2B1413
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 100800 (seconds)
    TLS session ticket:
    0000 - 5d a2 d5 29 fe 27 11 4e-31 77 56 f6 d2 80 ab 53   ]..).'.N1wV....S
    0010 - 6c 2c e0 97 8e bf c0 07-dd 66 74 6a 70 c3 f7 ca   l,.......ftjp...
    0020 - aa a8 20 b7 0a 9e 05 69-97 b2 84 b4 d3 1d fa 38   .. ....i.......8
    0030 - 3b 7f 17 f7 2a 4e 2e 3e-cc 53 08 49 bc ca 9b c7   ;...*N.>.S.I....
    0040 - 08 46 ec b0 e8 1d 52 97-34 4d 1d 6f f4 2f 7f 86   .F....R.4M.o./..
    0050 - 10 b5 0b 4f 85 73 c6 4c-ed 4b c0 77 c8 c3 45 d1   ...O.s.L.K.w..E.
    0060 - d6 35 a6 62 0a 41 4f b3-5e 33 e0 65 ec b4 19 e4   .5.b.AO.^3.e....
    0070 - ce a2 b1 62 dd 29 8e 2c-33 8b 5e 67 f1 03 f5 b0   ...b.).,3.^g....
    0080 - 9d 4a 0b b2 30 10 83 4d-ae ac 66 20 41 99 6a bf   .J..0..M..f A.j.
    0090 - ed b0 b1 34 66 76 92 63-7d 63 f0 a4 e1 4b 5d 77   ...4fv.c}c...K]w
    00a0 - ec 24 ed ab                                       .$..

    Start Time: 1467238146
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
...
```